Brittany Hanley
Week 10 Answers

create working directory and change into that directory
$mkdir week10
$cd week10

create and activate mamba environment
$mamba create -n week10 ddocent
$source activate week10

get data
$cp -r /home/BIO594/DATA/Week10/* .


samples already demultiplexed, skipping forward to create list of unique reads

$ls *.F.fq.gz > namelist #list all sample names and write into a namelist
$sed -i'' -e 's/.F.fq.gz//g' namelist #search for .F.fq.gz and replaces it with nothing
$AWK1='BEGIN{P=1}{if(P==1||P==2){gsub(/^[@]/,">");print}; if(P==4)P=0; P++}'
$AWK2='!/>/'
$AWK3='!/NNN/'
$PERLT='while (<>) {chomp; $z{$_}++;} while(($k,$v) = each(%z)) {print "$v\t$k\n";}'
 
$cat namelist | parallel --no-notice -j 8 "zcat {}.F.fq.gz | mawk '$AWK1' | mawk '$AWK2' > {}.forward"
$cat namelist | parallel --no-notice -j 8 "zcat {}.R.fq.gz | mawk '$AWK1' | mawk '$AWK2' > {}.reverse"
$cat namelist | parallel --no-notice -j 8 "paste -d '-' {}.forward {}.reverse | mawk '$AWK3' | sed 's/-/NNNNNNNNNN/' | perl -e '$PERLT' > {}.uniq.seqs"


eliminate reads with low copy numbers: sum numbers within individual coverage levels

$cat *.uniq.seqs > uniq.seqs
$for i in {2..20};
$do 
$echo $i >> pfile
$done
$cat pfile | parallel --no-notice "echo -n {}xxx && mawk -v x={} '\$1 >= x' uniq.seqs | wc -l" | mawk  '{gsub("xxx","\t",$0); print;}'| sort -g > uniqseq.data
$rm pfile


view unique reads
$more uniqseq.data

#Output
2       126508
3       122492
4       121026
5       119357
6       117552
7       115633
8       113606
9       111449
10      109143
11      106710
12      104198
13      101560
14      98723
15      95844
16      92812
17      89829
18      86728
19      83535
20      80401


plot the data using gnuplot
$gnuplot << \EOF
$set terminal dumb size 120, 30
$set autoscale
$set xrange [2:20]
$unset label
$set title "Number of Unique Sequences with More than X Coverage (Counted within $individuals)"
$set xlabel "Coverage"
$set ylabel "Number of Unique Sequences"
$plot 'uniqseq.data' with lines notitle
$pause -1
$EOF

#Output

                    Number of Unique Sequences with More than X Coverage (Counted within individuals)
 130000 +-------------------------------------------------------------------------------------------------------+
        |           +          +           +          +           +          +           +          +           |
 125000 |**                                                                                                   +-|
        |  ******                                                                                               |
        |        ******                                                                                         |
 120000 |-+            *****                                                                                  +-|
        |                   ******                                                                              |
 115000 |-+                       ******                                                                      +-|
        |                               ******                                                                  |
 110000 |-+                                   *****                                                           +-|
        |                                          ******                                                       |
 105000 |-+                                              ******                                               +-|
        |                                                      ******                                           |
        |                                                            ******                                     |
 100000 |-+                                                                ****                               +-|
        |                                                                      ***                              |
  95000 |-+                                                                       ****                        +-|
        |                                                                             ******                    |
  90000 |-+                                                                                 *****             +-|
        |                                                                                        **             |
        |                                                                                          ****         |
  85000 |-+                                                                                            *****  +-|
        |           +          +           +          +           +          +           +          +       *** |
  80000 +-------------------------------------------------------------------------------------------------------+
        2           4          6           8          10          12         14          16         18          20
                                                        Coverage
                                                        
                                                        
chose cutoff value of 4 to keep as much diversity in data while removing errors
$parallel --no-notice -j 8 mawk -v x=4 \''$1 >= x'\' ::: *.uniq.seqs | cut -f2 | perl -e 'while (<>) {chomp; $z{$_}++;} while(($k,$v) = each(%z)) {print "$v\t$k\n";}' > uniqCperindv


number of unique sequences
$wc -l uniqCperindv
#Output
12699 uniqCperindv


restrict data by numner of different individuals the sequence appears in
$for ((i = 2; i <= 10; i++));
$do
$echo $i >> ufile
$done

$cat ufile | parallel --no-notice "echo -n {}xxx && mawk -v x={} '\$1 >= x' uniqCperindv | wc -l" | mawk  '{gsub("xxx","\t",$0); print;}'| sort -g > uniqseq.peri.data
$rm ufile


plot restricted data
$gnuplot << \EOF
$set terminal dumb size 120, 30
$set autoscale
$unset label
$set title "Number of Unique Sequences present in more than X Individuals"
$set xlabel "Number of Individuals"
$set ylabel "Number of Unique Sequences"
$plot 'uniqseq.peri.data' with lines notitle
$pause -1
$EOF

#Output

                             Number of Unique Sequences present in more than X Individuals
 11000 +--------------------------------------------------------------------------------------------------------+
       |            +            +            +             +            +            +            +            |
       |                                                                                                        |
       |*                                                                                                       |
 10000 |-**                                                                                                   +-|
       |   ***                                                                                                  |
       |      ***                                                                                               |
       |         **                                                                                             |
  9000 |-+         ***                                                                                        +-|
       |              ***                                                                                       |
       |                 ****                                                                                   |
  8000 |-+                   ***                                                                              +-|
       |                        *****                                                                           |
       |                             ******                                                                     |
       |                                   *******                                                              |
  7000 |-+                                        *******                                                     +-|
       |                                                 *******                                                |
       |                                                        ******                                          |
       |                                                              *******                                   |
  6000 |-+                                                                   ******                           +-|
       |                                                                           **********                   |
       |                                                                                     **********         |
       |            +            +            +             +            +            +            +   ******   |
  5000 +--------------------------------------------------------------------------------------------------------+
       2            3            4            5             6            7            8            9            10
                                                 Number of Individuals


chose a new cutoff value of 4
$mawk -v x=4 '$1 >= x' uniqCperindv > uniq.k.4.c.4.seqs
wc -l uniq.k.4.c.4.seqs

#Output
7989 uniq.k.4.c.4.seqs


convert sequences to fasta
$cut -f2 uniq.k.4.c.4.seqs > totaluniqseq
$mawk '{c= c + 1; print ">Contig_" c "\n" $1}' totaluniqseq > uniq.fasta


assemble reference contigs

extract forward reads
$sed -e 's/NNNNNNNNNN/\t/g' uniq.fasta | cut -f1 > uniq.F.fasta

cluster forwards reads by 80% similarity
$cd-hit-est -i uniq.F.fasta -o xxx -c 0.8 -T 0 -M 0 -g 1

================================================================
                            Output
----------------------------------------------------------------
total number of CPUs in the system is 80
Actual number of CPUs to be used: 80

total seq: 7989
longest and shortest : 116 and 116
Total letters: 926724
Sequences have been sorted

Approximated minimal memory consumption:
Sequence        : 1M
Buffer          : 80 X 17M = 1414M
Table           : 2 X 16M = 33M
Miscellaneous   : 4M
Total           : 1454M

Table limit with the given memory limit:
Max number of representatives: 248606
Max number of word counting entries: 6715988

# comparing sequences from          0  to       7238
.......---------- new table with     1234 representatives
# comparing sequences from       7238  to       7989
....................---------- new table with        0 representatives

     7989  finished       1234  clusters
     

converted output of cd-hit to match first phase of rainbow
$mawk '{if ($1 ~ /Cl/) clus = clus + 1; else  print $3 "\t" clus}' xxx.clstr | sed 's/[>Contig_,...]//g' | sort -g -k1 > sort.contig.cluster.ids
$paste sort.contig.cluster.ids totaluniqseq > contig.cluster.totaluniqseq
$sort -k2,2 -g contig.cluster.totaluniqseq | sed -e 's/NNNNNNNNNN/\t/g' > rcluster


find number of clusters
$cut -f2 rcluster | uniq | wc -l
#Output
1234


split loci into alleles
$rainbow div -i rcluster -o rbdiv.out


adding -f to control min frequency of allele to divide into its own cluster
$rainbow div -i rcluster -o rbdiv.out -f 0.5 -K 10


merge clusters
$rainbow merge -o rbasm.out -a -i rbdiv.out -r 2


isolating optimal contigs for RAD sequencing
$cat rbasm.out <(echo "E") |sed 's/[0-9]*:[0-9]*://g' | mawk ' {
$if (NR == 1) e=$2;
$else if ($1 ~/E/ && lenp > len1) {c=c+1; print ">dDocent_Contig_" e "\n" seq2 "NNNNNNNNNN" seq1; seq1=0; seq2=0;lenp=0;e=$2;fclus=0;len1=0;freqp=0;lenf=0}
$else if ($1 ~/E/ && lenp <= len1) {c=c+1; print ">dDocent_Contig_" e "\n" seq1; $seq1=0; seq2=0;lenp=0;e=$2;fclus=0;len1=0;freqp=0;lenf=0}
$else if ($1 ~/C/) clus=$2;
$else if ($1 ~/L/) len=$2;
$else if ($1 ~/S/) seq=$2;
$else if ($1 ~/N/) freq=$2;
$else if ($1 ~/R/ && $0 ~/0/ && $0 !~/1/ && len > lenf) {seq1 = seq; fclus=clus;lenf=len}
$else if ($1 ~/R/ && $0 ~/0/ && $0 ~/1/) {seq1 = seq; fclus=clus; len1=len}
$else if ($1 ~/R/ && $0 ~!/0/ && freq > freqp && len >= lenp || $1 ~/R/ && $0 ~!/0/ && $freq == freqp && len > lenp) {seq2 = seq; lenp = len; freqp=freq}
}' > rainbow.fasta


align contigs and cluster based on similarity (90%)
$cd-hit-est -i rainbow.fasta -o referenceRC.fasta -M 0 -T 0 -c 0.9

 ================================================================
                            Output
----------------------------------------------------------------
total number of CPUs in the system is 80
Actual number of CPUs to be used: 80

total seq: 1235
longest and shortest : 373 and 233
Total letters: 310904
Sequences have been sorted

Approximated minimal memory consumption:
Sequence        : 0M
Buffer          : 80 X 17M = 1412M
Table           : 2 X 16M = 33M
Miscellaneous   : 4M
Total           : 1450M

Table limit with the given memory limit:
Max number of representatives: 248606
Max number of word counting entries: 10919214

# comparing sequences from          0  to         15
---------- new table with       15 representatives
# comparing sequences from         15  to         29
..............---------- new table with       14 representatives
# comparing sequences from         29  to         43
..............---------- new table with       14 representatives
# comparing sequences from         43  to         57
..............---------- new table with       14 representatives
# comparing sequences from         57  to         71
..............---------- new table with       14 representatives
# comparing sequences from         71  to         85
..............---------- new table with       14 representatives
# comparing sequences from         85  to         99
..............---------- new table with       14 representatives
# comparing sequences from         99  to        112
.............---------- new table with       13 representatives
# comparing sequences from        112  to        125
.............---------- new table with       13 representatives
# comparing sequences from        125  to        138
.............---------- new table with       13 representatives
# comparing sequences from        138  to        151
.............---------- new table with       13 representatives
# comparing sequences from        151  to        164
.............---------- new table with       13 representatives
# comparing sequences from        164  to        177
.............---------- new table with       13 representatives
# comparing sequences from        177  to        189
............---------- new table with       12 representatives
# comparing sequences from        189  to        201
............---------- new table with       12 representatives
# comparing sequences from        201  to        213
............---------- new table with       12 representatives
# comparing sequences from        213  to        225
............---------- new table with       12 representatives
# comparing sequences from        225  to        237
............---------- new table with       12 representatives
# comparing sequences from        237  to       1235
.....................---------- new table with      998 representatives

     1235  finished       1235  clusters
     
     
remake ref using above parameters
$bash remake_reference.sh 4 4 0.90 PE 2


exmaine references
$head reference.fasta

#Output
>dDocent_Contig_1
NAATTCATGGGATTCCCTGAGAGCACGAACGTCATTTACATCTAATACTCATTGGCACGTCATGCATTGGCAAAGACGAGTAGTTAGTGATAACGCCTAATCACCACGTCAACTGAANNNNNNNNNNGTGGGGTAGCGGGATAGTAAGCCCCCTGGATTGTCCTATTGACTGCGAAGACAAATAGCAGAGGTTCATACGCTCGGTCCTTTGCGCAGAGGACGGACGATTCGGACGCTATCCTAATTTTGCCGN
>dDocent_Contig_2
NAATTCGTTTGCCCACGGCTTCCACTAAAAGTTGCCCGCAGAACGGATCACTCCAGTATATGCTGCAGTTTGGATGTGAGAGCGGATAGTTTTGCTAATCATCCACGGGCCGTTAGTNNNNNNNNNNCGTAAGGCATCGGATTTACACGGTATACGTCCGATGCATTGCTTGTACCCACGTCCGAATTCATCGACGTGCGCACTCCTGATTATAACTTAACAATAGTCATAACGCCGGGCTCCCTGGTTCCGN
>dDocent_Contig_3
NAATTCACGGCTATCAACTAGGATGGTGGTTACTATTAGTGAGTGCTGTGTATTTCCGCTGCCGTCACTTGCAAGGCAGTAAACCCTTGGTGGCACGTGTAATCCAGCGTATGCAATNNNNNNNNNNCCTGGCCATAGTATTGCTCTAGCATAAAACAAGAGTTATGTATCTTGCCTTCCGGCTAGTCACCTATAGTGATTTGAGCTATTGAAAAGTCACGTTGACTGGAGGTAGAGAGTGGAATACTTCCGN
>dDocent_Contig_4
NAATTCAGCAACTCAAGTAATTCTGTGACTGCCACACCTTTCACCTGTAAGGCACTCGCGTACATCATTAGATCTTATTTGAAAGACCTGGCGTCGCCAATGTTGTCCGCAATAATCNNNNNNNNNNTGCGTACTCACGTTGTTATATAATGCAGCCGTCACACAATTTCGTGGATCGGCTACGGTGCGGGACTGAGACATACGTACGGTCAATAGGAGTAATAATCGCTTCATCATGATACTGGCTGTCCGN
>dDocent_Contig_5
NAATTCTCTGGCATTAATACCTTTATTTCTTTCCCGGAATTTGGTCCATACGCCAAAACCACATTAACTTTACACAGACCATGTCCTCGACGGTCGAATTTAGACAAATTTCTAGTGNNNNNNNNNNCCGTTATTGAACCGAACTATCCTGTTTGATTCGGGGCCTTGGATTTTGACTGGCGTAAGTGCACCGAATTTATAGTATACAATTTTTCACGGGGTAGACGAGTGCGATATCGATCGAGTGAACCGN


counting lines in file
$wc -l reference.fasta
#Output
2470 reference.fasta


counting number of contigs in file
$mawk '/>/' reference.fasta | wc -l
#Output
1235

test that parameters were followed
$mawk '/^NAATT.*N*.*CCGN$/' reference.fasta | wc -l
$grep '^NAATT.*N*.*CCGN$' reference.fasta | wc -l
#Output
1235

