---
title: "Final Project for BIO 594"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## GitHub Documents

This is an R Markdown format used for publishing markdown documents to GitHub. When you click the **Knit** button all R code chunks are run and a markdown file (.md) suitable for publishing to GitHub is generated.

# Link to original github: https://github.com/pdimens/2022-Tatlanticus_popgen

# Data setup
* Population: 12 populations from 5 different localities (10 near sewage effluent sources, 2 controls)
* The sequenced data files are located on KITT and can be accessed via

  PATH:  `/home/BIO594/Final_Project`
* Size: 376 individuals (28-33 per population), 150bp sequences
* Sequences were previously demultiplexed with barcodes and individuals from a different species were removed (completed by Dr. Jon Puritz)

# Bioinformatics

## Initial Raw Data Assessment and Characterization (Jon)
  * Count the RAW reads
  * Examine quality of data (fastqc)

## Reference assembly (Cass)
   * De Novo Reference Assembly 
      * Include optimization
      * Assemble only a subset of individuals

## RADSeq Bioinformatics (Caitlin)
   * Trimming adapters and low quality reads (fastp via dDocent)
   * Read Mapping (BWA & SAMtools via dDocent)
      * Try using different mappaing parameters on a subset and pick the best ones
   * SNP calling and filtering (FreeBayes via dDocent)

## SNP Filtering (VCFtools, vcflib, rad_haplotyper) (Willow)

# Population-level Analyses

## Detecting Selection Via Outlier Detection Programs (Flow and Michelle)
      * RDA 
      * Bayenv2
      * PCAadapt
      * Outflank

## Detecting Population Structure (Brittany and Angela)
      * Pairwise *F~ST~*
      * Genetic diversity (observed and expected heterozygosity)
      * EEMS (https://www.nature.com/articles/ng.3464)






This data set contains technical replicates. You can use the custom script `dup_sample_filter.sh` to automatically remove sites in VCF files that do not have congruent genotypes across duplicate individuals. It will only consider genotypes that have at least 5 reads.


The technical replicates are listed in `duplicates.samples` with the following format (each name should be separated by a tab):
```bash
FBN_327a        FBN_327b
FBN_327a        FBN_327c
FBN_327a        FBN_327d
FBN_327b        FBN_327c
FBN_327b        FBN_327d
FBN_327c        FBN_327d
OBN_9b         OBN_9c
OBN_9b         OBN_9d
OBN_9c         OBN_9d
OBS_245a        OBS_245b
OBS_245a        OBS_245c
OBS_245a        OBS_245d
OBS_245b        OBS_245c
OBS_245b        OBS_245d
OBS_245c        OBS_245d
PCN_210a        PCN_210b
PCN_223a        PCN_223b
PCS_361a        PCS_361b
PCS_365a        PCS_365b
SPS_74a        SPS_74b
SPS_92a        SPS_92b
SPS_92a        SPS_92c
SPS_92a        SPS_92d
SPS_92b        SPS_92c
SPS_92b        SPS_92d
SPS_92c        SPS_92d
WC2_301a        WC2_301b
WC2_305a        WC2_305b
WC2_305a        WC2_305c
WC2_305b        WC2_305c
```  

Run script.
```bash
./dup_sample_filter.sh Filtered.SNP.VCF duplicates.samples
```
This produces a `mismatched.loci` file.
```bash
head mismatched.loci
```
```bash
2       dDocent_Contig_6167     14
2       dDocent_Contig_12111    122
2       dDocent_Contig_5485     184
2       dDocent_Contig_11981    240
14      dDocent_Contig_5512     27
2       dDocent_Contig_11857    114
2       dDocent_Contig_11642    216
2       dDocent_Contig_3841     25
2       dDocent_Contig_11322    100
2       dDocent_Contig_3530     210
```

```bash
echo -e "Mismatches\tNumber_of_Loci" > mismatch.txt
for i in {2..20}
do
paste <(echo $i) <(mawk -v x=$i '$1 > x' mismatched.loci | wc -l) >> mismatch.txt
done
```
In RStudio
```r
library(ggplot2)
mismatch <- read.table("mismatch.txt", header = TRUE)
df=data.frame(mismatch)

p <- ggplot(df, aes(x=Mismatches, y=Number_of_Loci)) + geom_point() +theme_bw() + scale_x_continuous(minor_breaks = seq(1,20,by=1))
p
```
![mismatched](https://github.com/amaeliazyck/RADseq_Uca-rapax_2016/blob/master/Output/Filtering/mismatched.png)

I decided to filter out loci with mismatched values > 6.
```bash
mawk '$1 > 6' mismatched.loci | cut -f2,3 > mismatchedloci

vcftools --vcf SNP.TRSdp5g5mafMIap9g9d.recode.vcf --exclude-positions mismatchedloci --recode --recode-INFO-all --out SNP.TRSdp5g5mafMIap9g9dMM
```
```bash
After filtering, kept 340 out of 340 Individuals
Outputting VCF file...
After filtering, kept 43532 out of a possible 43983 Sites
Run Time = 97.00 seconds
```





