---
title: "Final_Project_PCAdapt_Outflank_Outlier_Detection"
author: "Flo"
date: '2024-05-08'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

 
```{r}
#Load Libraries
library(vcfR)
library(pcadapt)
library(OutFLANK) 
library(vcfR)
library(bigsnpr)
library(qvalue)
library(dplyr)
```

PCAdapt
```{r}
filteredSNP <- read.pcadapt("SNP.TRSdp5g5mac3mafMIasubp9g9dMMHWEmaf0252A.recode.vcf", type = "vcf") 
x <- pcadapt(input = filteredSNP, K = 20) 
plot(x,option = "screeplot")
plot(x, option = "screeplot", K = 10) 
```
```{bash, eval=FALSE}
#Run in terminal
cut -f1 out.imiss |grep -v INDV| cut -f1 -d "_" | sort | uniq -c
#output
#30 FBN
#25 FBS
#19 OBN
#28 OBS
#27 PCN
#25 PCS
#26 SPN
#29 SPS
#25 WC1
#27 WC2
#28 WC3
#29 WC4
```

```{r}

poplist.names <- c(rep("FBN", 30),rep("FBS", 25),rep("OBN", 19), rep("OBS",28), rep("PCN",27), rep("PCS",25), rep("SPN",26), rep("SPS",29), rep("WC1",25), rep("WC2",27), rep("WC3",28), rep("WC4",29))


plot(x, option = "scores", pop = poplist.names) 
plot(x, option = "scores", i = 2, j = 3, pop = poplist.names)
plot(x, option = "scores", i = 3, j = 4, pop = poplist.names) 
```
Choosing 3 to be the cutoff

```{r}
x <- pcadapt(filteredSNP, K = 3)
summary(x)

plot(x , option = "manhattan") 
plot(x, option = "qqplot", threshold = 0.1) 
plot(x, option = "stat.distribution") 
```

```{r}
# Set FDR
qval1 <- qvalue(x$pvalues)$qvalues
alpha <- 0.1
```

```{r}
# Save outliers
outliers1 <- which(qval1 < alpha)
```

```{r}
invisible(lapply(outliers1, write, "outliers1.txt", append=TRUE))
```

```{bash eval= FALSE}
#Run in terminal
mawk '!/#/' SNP.TRSdp5g5mac3mafMIasubp9g9dMMHWEmaf0252A.recode.vcf | cut -f1,2 > totalloci
NUM=(`cat totalloci | wc -l`)
paste <(seq 1 $NUM) totalloci > loci.plus.index
cat outliers1.txt | parallel "grep -w ^{} loci.plus.index" | cut -f2,3> outliers1.l.txt
```


# OutFlank
```{r}
filteredSNP <- read.vcfR("SNP.TRSdp5g5mac3mafMIasubp9g9dMMHWEmaf0252A.recode.vcf")
```

```{r}
geno <- extract.gt(filteredSNP) 
position <- getPOS(filteredSNP) 
chromosome <- getCHROM(filteredSNP)
```

```{r}
G <- matrix(NA, nrow = nrow(geno), ncol = ncol(geno))

G[geno %in% c("0/0", "0|0")] <- 0
G[geno  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
G[geno %in% c("1/1", "1|1")] <- 2

G[is.na(G)] <- 9

head(G[,1:10])
```

#No need to use popmap if you have pop.list
```{r}
pop <- as.vector(poplist.names)
```

```{r}
my_fst <- MakeDiploidFSTMat(t(G), locusNames = paste0(chromosome,"_", position), popNames = pop)
plot(my_fst$He, my_fst$FST)
my_dist <- OutFLANK(my_fst, NumberOfSamples = 12, qthreshold=0.1, RightTrimFraction=0.1, LeftTrimFraction=0.1)
```

```{r}

OutFLANKResultsPlotter(my_dist)

plot(my_dist$results$FST, col=as.numeric(as.factor(chromosome)))
my_dist$results[which(my_dist$results$OutlierFlag == TRUE),]
```


```{bash eval = FALSE}
#Run in terminal
mawk '!/#/' SNP.TRSdp5g5mac3mafMIasubp9g9dMMHWEmaf0252A.recode.vcf | cut -f1,2 > totalloci
NUM=(`cat totalloci | wc -l`)
paste <(seq 1 $NUM) totalloci > loci1.plus.index
echo -e "3209"| parallel "grep -w ^{} loci.plus.index" | cut -f2,3> outliers2.l.txt
echo -e "3210"| parallel "grep -w ^{} loci.plus.index" | cut -f2,3> outliers3.l.txt
echo -e "5704"| parallel "grep -w ^{} loci.plus.index" | cut -f2,3> outliers4.l.txt
echo -e "6463"| parallel "grep -w ^{} loci.plus.index" | cut -f2,3> outliers5.l.txt
echo -e "6464"| parallel "grep -w ^{} loci.plus.index" | cut -f2,3> outliers6.l.txt
echo -e "7551"| parallel "grep -w ^{} loci.plus.index" | cut -f2,3> outliers7.l.txt
```
